File: packages/MeeshySDK/Sources/MeeshySDK/Services/ReportService.swift
import Foundation

// MARK: - Protocol

public protocol ReportServiceProviding: Sendable {
    func reportMessage(messageId: String, reportType: String, reason: String?) async throws
    func reportUser(userId: String, reportType: String, reason: String?) async throws
    func reportPost(postId: String, reportType: String, reason: String?) async throws
    func reportStory(storyId: String, reportType: String, reason: String?) async throws
    func reportConversation(conversationId: String, reportType: String, reason: String?) async throws
}

/// Signaler un contenu est un geste ORDINAIRE (S2), pas un geste
/// d'administration (#4155).
///
/// Ces cinq appels visaient `/admin/reports` — la seule route d'administration
/// que l'app appelait. Toute règle d'infrastructure posée un jour sur le
/// préfixe `/admin` (liste blanche d'IP, WAF, journalisation renforcée) aurait
/// cassé le signalement sur les trois plateformes, sans que personne ne fasse
/// le lien. L'adresse dit désormais ce que le geste est.
///
/// `/admin/reports` reste servie en adaptateur mince le temps que les versions
/// déjà installées disparaissent.
public final class ReportService: ReportServiceProviding, @unchecked Sendable {
    public static let shared = ReportService()
    private let api: APIClientProviding

    init(api: APIClientProviding = APIClient.shared) {
        self.api = api
    }

    public func reportMessage(messageId: String, reportType: String, reason: String? = nil) async throws {
        let body = CreateReportBody(
            reportedType: "message",
            reportedEntityId: messageId,
            reportType: reportType,
            reason: reason
        )
        let _: APIResponse<ReportResponseData> = try await api.post(endpoint: "/reports", body: body)
    }

    public func reportUser(userId: String, reportType: String, reason: String? = nil) async throws {
        let body = CreateReportBody(
            reportedType: "user",
            reportedEntityId: userId,
            reportType: reportType,
            reason: reason
        )
        let _: APIResponse<ReportResponseData> = try await api.post(endpoint: "/reports", body: body)
    }

    public func reportPost(postId: String, reportType: String, reason: String? = nil) async throws {
        let body = CreateReportBody(
            reportedType: "post",
            reportedEntityId: postId,
            reportType: reportType,
            reason: reason
        )
        let _: APIResponse<ReportResponseData> = try await api.post(endpoint: "/reports", body: body)
    }

    public func reportStory(storyId: String, reportType: String, reason: String? = nil) async throws {
        let body = CreateReportBody(
            reportedType: "story",
            reportedEntityId: storyId,
            reportType: reportType,
            reason: reason
        )
        let _: APIResponse<ReportResponseData> = try await api.post(endpoint: "/reports", body: body)
    }

    public func reportConversation(conversationId: String, reportType: String, reason: String? = nil) async throws {
        let body = CreateReportBody(
            reportedType: "conversation",
            reportedEntityId: conversationId,
            reportType: reportType,
            reason: reason
        )
        let _: APIResponse<ReportResponseData> = try await api.post(endpoint: "/reports", body: body)
    }
}

struct CreateReportBody: Encodable {
    let reportedType: String
    let reportedEntityId: String
    let reportType: String
    let reason: String?
}

public struct ReportResponseData: Decodable {
    public let id: String?
}

File: apps/web/services/report.service.ts
import { apiService } from './api.service';
import type { CreateReportDTO, Report } from '@meeshy/shared/types';

/**
 * Service pour gérer les signalements.
 *
 * Adresse `POST /reports` depuis #4155 : signaler est un geste ORDINAIRE (S2).
 * Il vivait sous `/admin/reports`, la seule route d'administration ouverte à un
 * utilisateur ordinaire — une adresse qui mentait sur le privilège, et que
 * n'importe quel durcissement du préfixe `/admin` aurait cassée.
 */
class ReportService {
  /**
   * Signaler un message
   */
  async reportMessage(messageId: string, reportType: string, reason: string): Promise<Report> {
    const reportData: CreateReportDTO = {
      reportedType: 'message',
      reportedEntityId: messageId,
      reportType,
      reason
    };

    const response = await apiService.post<{ success: boolean; data: Report }>('/reports', reportData);

    if (response.data && (response.data as any).success) {
      return (response.data as any).data;
    }

    return response.data?.data as Report;
  }

  /**
   * Signaler un utilisateur
   */
  async reportUser(userId: string, reportType: string, reason: string): Promise<Report> {
    const reportData: CreateReportDTO = {
      reportedType: 'user',
      reportedEntityId: userId,
      reportType,
      reason
    };

    const response = await apiService.post<{ success: boolean; data: Report }>('/reports', reportData);

    if (response.data && (response.data as any).success) {
      return (response.data as any).data;
    }

    return response.data?.data as Report;
  }

  /**
   * Signaler une conversation
   */
  async reportConversation(conversationId: string, reportType: string, reason: string): Promise<Report> {
    const reportData: CreateReportDTO = {
      reportedType: 'conversation',
      reportedEntityId: conversationId,
      reportType,
      reason
    };

    const response = await apiService.post<{ success: boolean; data: Report }>('/reports', reportData);

    if (response.data && (response.data as any).success) {
      return (response.data as any).data;
    }

    return response.data?.data as Report;
  }

  /**
   * Signaler une communauté
   */
  async reportCommunity(communityId: string, reportType: string, reason: string): Promise<Report> {
    const reportData: CreateReportDTO = {
      reportedType: 'community',
      reportedEntityId: communityId,
      reportType,
      reason
    };

    const response = await apiService.post<{ success: boolean; data: Report }>('/reports', reportData);

    if (response.data && (response.data as any).success) {
      return (response.data as any).data;
    }

    return response.data?.data as Report;
  }

  /**
   * Signaler un post
   */
  async reportPost(postId: string, reportType: string, reason: string): Promise<Report> {
    const reportData: CreateReportDTO = {
      reportedType: 'post',
      reportedEntityId: postId,
      reportType,
      reason
    };

    const response = await apiService.post<{ success: boolean; data: Report }>('/reports', reportData);

    if (response.data && (response.data as any).success) {
      return (response.data as any).data;
    }

    return response.data?.data as Report;
  }

  /**
   * Signaler une story
   */
  async reportStory(storyId: string, reportType: string, reason: string): Promise<Report> {
    const reportData: CreateReportDTO = {
      reportedType: 'story',
      reportedEntityId: storyId,
      reportType,
      reason
    };

    const response = await apiService.post<{ success: boolean; data: Report }>('/reports', reportData);

    if (response.data && (response.data as any).success) {
      return (response.data as any).data;
    }

    return response.data?.data as Report;
  }
}

// Instance singleton
export const reportService = new ReportService();
